lib/service/modules/entity.js

```diff
@@ -0,0 +1,185 @@
+    let _entity = {};
+
+    const _createEntityId = () => {
+      if (useBSONObjectId) return new BSON.ObjectID();
+      return new BSON.ObjectID().toString();
+    };
+
+    const currentTimeString = moment().utc().toISOString();
+
+    if (typeof entityOrId === 'string' || typeof entityOrId === 'undefined') {
+      _entity = {
+        _acl: {
+          creator: environmentId
+        },
+        _kmd: {
+          lmt: currentTimeString,
+          ect: currentTimeString
+        }
+      };
+
+      if (entityOrId) {
+        _entity._id = entityOrId;
+      } else {
+        _entity._id = _createEntityId();
+      }
+    } else {
+      if (typeof entityOrId === 'object') {
+        _entity = entityOrId;
+
+        if (!_entity._acl) _entity._acl = {};
+        if (!_entity._id) _entity._id = _createEntityId();
+        if (!_entity._acl) _entity._acl = {};
+        if (!_entity._acl.creator) _entity._acl.creator = environmentId;
+        if (!_entity._kmd) _entity._kmd = {};
+        if (!_entity._kmd.ect) _entity._kmd.ect = currentTimeString;
+        if (!_entity._kmd.lmt) _entity._kmd.lmt = currentTimeString;
+      }
+    }
+
+    const aclFunctions = {
+      getCreator: () => _.get(_entity, ['_acl', 'creator']),
+      getReaders: () => _.get(_entity, ['_acl', 'r'], []),
+      getWriters: () => _.get(_entity, ['_acl', 'w'], []),
+      getReaderGroups: () => _.get(_entity, ['_acl', 'groups', 'r'], []),
+      getWriterGroups: () => _.get(_entity, ['_acl', 'groups', 'w'], []),
+      addReader: (userId) => {
+        if (!(_.get(_entity, ['_acl', 'r']))) _entity._acl.r = [];
+        if (_entity._acl.r.indexOf(userId) === -1) _entity._acl.r.push(userId);
+        return aclFunctions;
+      },
+      addWriter: (userId) => {
+        if (!(_.get(_entity, ['_acl', 'w']))) _entity._acl.w = [];
+        if (_entity._acl.w.indexOf(userId) === -1) _entity._acl.w.push(userId);
+        return this;
+      },
+      addReaderGroup: (groupId) => {
+        if (!(_.get(_entity, ['_acl', 'groups']))) _entity._acl.groups = [];
+        if (!(_.get(_entity, ['_acl', 'groups', 'r']))) _entity._acl.groups.r = [];
+        if (_entity._acl.groups.r.indexOf(groupId) === -1) _entity._acl.groups.r.push(groupId);
+        return aclFunctions;
+      },
+
+      addWriterGroup: (groupId) => {
+        if (!(_.get(_entity, ['_acl', 'groups']))) _entity._acl.groups = [];
+        if (!(_.get(_entity, ['_acl', 'groups', 'w']))) _entity._acl.groups.w = [];
+        if (_entity._acl.groups.w.indexOf(groupId) === -1) _entity._acl.groups.w.push(groupId);
+        return aclFunctions;
+      },
+      removeReader: (userId) => {
+        let readers = _.get(_entity, ['_acl', 'r'], []);
+        let userIx = readers.indexOf(userId);
+        while (userIx !== -1) {
+          _entity._acl.r.splice(userIx, 1);
+          readers = _.get(_entity, ['_acl', 'r'], []);
+          userIx = readers.indexOf(userId);
```

Comment:
Why not just `userIx = _entity._acl.r.indexOf(userId)`? by this point you know that the path exists and is an array.